### ISSUE TITLE
[thread 1 callback] bad argument #2 to '?' for own dataset

### ISSUE THREAD

**Issue by zelle-github** [facebookresearch/deepmask#72]
Hi,

I try to train on my own dataset and am following the steps of "Training your own model" of the github page. I was able to train with the coco dataset but run into problems using my own data. I have segmentations in a json file and successfully converted them to t7 files. But when I run th train.lua I get the following errors:

`
:/facebookResearch$ th train.lua -nthreads 1 -batch 4 -maxload 500
batch	4	32
-- ignore option rundir	
-- ignore option dm	
-- ignore option reload	
-- ignore option gpu	
-- ignore option datadir	
nthreads	1	2	
maxload	500	4000	
| running in directory /facebookResearch/exps/deepmask/exp,batch=4,maxload=500,nthreads=1	
| number of paramaters trunk: 15213312	
| number of paramaters mask branch: 1608768

| number of paramaters score branch: 526337	
| number of paramaters total: 17348417	
| start training	
/torch/install/bin/luajit: /torch/install/share/lua/5.1/threads/threads.lua:183: [thread 1 callback] bad argument #2 to '?' (out of range at /home/zelle/torch/pkg/torch/generic/Tensor.c:890)
stack traceback:
	[C]: at 0x7fcc88aa06a0
	[C]: in function '__index'
	/facebookResearch/DataSampler.lua:135: in function 'maskSampling'
	/facebookResearch/DataSampler.lua:111: in function 'get'
	/facebookResearch/DataLoader.lua:70: in function </facebookResearch/DataLoader.lua:64>
	[C]: in function 'xpcall'
	/torch/install/share/lua/5.1/threads/threads.lua:234: in function 'callback'
	/torch/install/share/lua/5.1/threads/queue.lua:65: in function </torch/install/share/lua/5.1/threads/queue.lua:41>
	[C]: in function 'pcall'
	/torch/install/share/lua/5.1/threads/queue.lua:40: in function 'dojob'
	[string "  local Queue = require 'threads.queue'..."]:15: in main chunk
stack traceback:
	[C]: in function 'error'
	/torch/install/share/lua/5.1/threads/threads.lua:183: in function 'dojob'
	/facebookResearch/DataLoader.lua:95: in function '(for generator)'
	/facebookResearch/TrainerDeepMask.lua:68: in function 'train'
	train.lua:117: in main chunk
	[C]: in function 'dofile'
	...torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x560b40529450
`
I have the current version of code from github and installed nn, nnx, inn, cunn.

Is it a problem with my data? It works with the coco data. But conversion of my json file to t7 worked fine.

Thanks for helping

**Comment by PavlosMelissinos** [comment]
No, it's not a problem with your json file.

[Line 135 in DataSampler](https://github.com/facebookresearch/deepmask/blob/master/DataSampler.lua#L135) (function maskSampling) is the following:

> local catId = self.catIds[cat]

The error is:

> (out of range at /home/zelle/torch/pkg/torch/generic/Tensor.c:890)

Print the size of the array (self.catIds) right before that line, as well as cat, and you'll see that the former is smaller than the latter.

The problem is that the number of categories is **hardcoded** to be 80, because the code only works for MS-COCO. Your dataset obviously has fewer classes than that; therefore you have to change [line 132](https://github.com/facebookresearch/deepmask/blob/master/DataSampler.lua#L132):

> local cat,ann = torch.random(80)

to represent the number of categories in your dataset.

**Comment by zelle-github** [comment]
Thats it. Thank you very much! GPU and CPU are having fun now :-)

**Comment by PavlosMelissinos** [comment]
Great, glad I could help! If the issue has been solved indeed, you should close it.

**Comment by PavlosMelissinos** [comment]
Actually, it seems like this is a duplicate of #42 .
